File: src/leetcode/medium/design-add-and-search-words-data-structure/design_add_and_search_words_data_structure.py
"""
https://leetcode.com/problems/design-add-and-search-words-data-structure

Design a data structure that supports adding new words and finding if a string matches any previously added string.

Implement the WordDictionary class:

WordDictionary() Initializes the object.
void addWord(word) Adds word to the data structure, it can be matched later.
bool search(word) Returns true if there is any string in the data structure that matches word or false otherwise. word may contain dots '.' where dots can be matched with any letter.

Example:
Input
["WordDictionary","addWord","addWord","addWord","search","search","search","search"]
[[],["bad"],["dad"],["mad"],["pad"],["bad"],[".ad"],["b.."]]
Output
[null,null,null,null,false,true,true,true]

Explanation
WordDictionary wordDictionary = new WordDictionary();
wordDictionary.addWord("bad");
wordDictionary.addWord("dad");
wordDictionary.addWord("mad");
wordDictionary.search("pad"); // return False
wordDictionary.search("bad"); // return True
wordDictionary.search(".ad"); // return True
wordDictionary.search("b.."); // return True

Constraints:
1 <= word.length <= 25
word in addWord consists of lowercase English letters.
word in search consist of '.' or lowercase English letters.
There will be at most 3 dots in word for search queries.
At most 104 calls will be made to addWord and search.
"""

class Node:
    def __init__(self, word_totals: int = 0, value = None):
        self.value = value
        self.children = {}
        self.word_totals = word_totals

class WordDictionary:
    def __init__(self):
        self.root = Node()
        

    def addWord(self, word: str) -> None:
        current = self.root
        word_index = 0
        
        while word_index < len(word):
            if not word[word_index] in current.children:
                current.children[word[word_index]] = Node(0, word[word_index])
                
            current = current.children[word[word_index]]
            word_index += 1
            
        current.word_totals += 1

    def search(self, word: str) -> bool:
        def dfs(j, root):
            cur = root
            
            for i in range(j, len(word)):
                c = word[i]
                if c == ".":
                    for child in cur.children.values():
                        if dfs(i + 1, child):
                            return True
                    return False
                else:
                    if c not in cur.children:
                        return False
                    cur = cur.children[c]
            return cur.word_totals > 0
        
        return dfs(0, self.root)

File: src/leetcode/medium/design-add-and-search-words-data-structure/designAddAndSearchWordsDataStructure.js
/*
https://leetcode.com/problems/design-add-and-search-words-data-structure

Design a data structure that supports adding new words and searching for existing words.

Implement the WordDictionary class:

void addWord(word) Adds word to the data structure.
bool search(word) Returns true if there is any string in the data structure that matches word or false otherwise. word may contain dots '.' where dots can be matched with any letter.

Example 1:
Input:
["WordDictionary", "addWord", "day", "addWord", "bay", "addWord", "may", "search", "say", "search", "day", "search", ".ay", "search", "b.."]
Output:
[null, null, null, null, false, true, true, true]

Explanation:
WordDictionary wordDictionary = new WordDictionary();
wordDictionary.addWord("day");
wordDictionary.addWord("bay");
wordDictionary.addWord("may");
wordDictionary.search("say"); // return false
wordDictionary.search("day"); // return true
wordDictionary.search(".ay"); // return true
wordDictionary.search("b.."); // return true

Constraints:
1 <= word.length <= 20
word in addWord consists of lowercase English letters.
word in search consist of '.' or lowercase English letters.
There will be at most 2 dots in word for search queries.
At most 10,000 calls will be made to addWord and search.
*/

class TrieNode {
    constructor(val) {
        this.val = val;
        this.isWord = false;
        this.children = {};
    }

    setAsWord() {
        this.isWord = true;
    }

    setChild(node) {
        this.children[node.val] = node;
    }

    getChild(val) {
        return this.children[val];
    }

    getChildren() {
        return this.children;
    }
}

class WordDictionary {
    constructor() {
        this.root = new TrieNode("__root__");
    }

    /**
     * @param {string} word
     * @return {void}
     */
    addWord(word) {
        let currentNode = this.root;
        for (let i = 0; i < word.length; i++) {
            const char = word[i];

            if (!currentNode.getChild(char)) {
                currentNode.setChild(new TrieNode(char));
            }
            currentNode = currentNode.getChild(char);

            if (i === word.length - 1) {
                currentNode.setAsWord();
            }
        }
    }

    /**
     * @param {string} word
     * @return {boolean}
     */
    search(word) {
        function bfs(node, i) {
            if (i === word.length && node.isWord) {
                return true;
            }

            const nodeChildren = node.getChildren();
            if (!Object.keys(nodeChildren).length) {
                return false;
            }

            const char = word[i];
            const results = [];
            if (char === ".") {
                for (const child of Object.keys(nodeChildren)) {
                    results.push(bfs(nodeChildren[child], i + 1));
                }
            } else if (nodeChildren[char]) {
                results.push(bfs(nodeChildren[char], i + 1));
            }

            return results.some((result) => result);
        }

        return bfs(this.root, 0);
    }
}

module.exports = { TrieNode, WordDictionary };
